### PR TITLE
in `flutter run`, throw tool exit when `--flavor` is provided but is not supported on the target device

### LINKED ISSUE #134197: `flutter run -d chrome --flavor <flavor>` runs without error but `flutter build web --flavor` results in tool exiting
Flavors are only an officially supported feature for Android, iOS, and (indirectly) MacOS. Running `flutter build web --flavor strawberry` will result in the tool exiting with this message:

```
Could not find an option named "flavor".

Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options.
```

However, one can run `flutter run -d chrome --flavor strawberry`, and the app will run without issue. At best, this is potentially confusing. At worst, it could result in Flutter accidentally supporting flavors for web if/when we add more flavors-related features to the framework.

We should also exit the tool when the user tries to use `--flavor` with `flutter run -d chrome`.

I recommend resolving this before implementing #21682

### PR DESCRIPTION
**Pull request by andrewkolos**
Fixes https://github.com/flutter/flutter/issues/134197

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -576,6 +576,20 @@ class RunCommand extends RunCommandBase {
     webMode = featureFlags.isWebEnabled &&
       devices!.length == 1  &&
       await devices!.single.targetPlatform == TargetPlatform.web_javascript;
+
+    final String? flavor = stringArg('flavor');
+    if (flavor != null) {
+      const List<PlatformType> platformsWithFlavorsSupport = <PlatformType>[
+        PlatformType.android,
+        PlatformType.ios,
+        PlatformType.macos,
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/commands/run.dart:585]: Can each device type instead declare whether it supports flavors?

https://github.com/flutter/flutter/blob/995a020e48aa6627ea3f904a59dd56349a819eb4/packages/flutter_tools/lib/src/device.dart#L759-L773

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/run.dart:585]: +1

**Comment by andrewkolos** [comment]
fyi @christopherfujino, this is now ready for review again (assuming my last commit doesn't break test)

File: packages/flutter_tools/lib/src/commands/run.dart (hunk 1 above)
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart]: Great idea! Done.

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 2]
```diff
@@ -576,6 +576,13 @@ class RunCommand extends RunCommandBase {
     webMode = featureFlags.isWebEnabled &&
       devices!.length == 1  &&
       await devices!.single.targetPlatform == TargetPlatform.web_javascript;
+
+    final String? flavor = stringArg('flavor');
+    final bool flavorsSupportedOnEveryDevice = devices!
+      .every((Device device) => device.supportsFlavors);
+    if (flavor != null && !flavorsSupportedOnEveryDevice) {
+      throwToolExit('--flavor is only supported for Android and iOS devices.');
```
[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: Does it work with `flutter run -d macos`? If so, fixing `flutter build macos` might be a one line fix?

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: Then we should document flavors support for macOS (assuming it works). IIRC the setup process in Xcode would be similar but not identical. I'll do more research.

[the remark this answers is not in the document]
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: Looks like we deleted the test in april. Not sure what we should do moving forward... https://github.com/flutter/flutter/pull/125581

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: If 

1. macOS support is not documented,
2. macOS support is not under test, and most importantly
3. users can't even build with flavors on macOS (unless they are publishing builds produced by `flutter run`),

then it's clear to me that macOS is not supported. If support is desired, then it should go through the normal feature request process, no?

[the remark this answers is not in the document]
Comment (jmagman) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: https://github.com/flutter/flutter/pull/119564
https://github.com/flutter/flutter/issues/64088
cc @cbracken

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: Chatted with @christopherfujino. I still think it's fair for this PR to merge as is, but it's arguably worth seeing if macOS flavor support can be easily fixed before making it officially unsupported.

I will see what the current state of macOS flavor support is (i.e. does `flutter run --flavor` even work properly for macOS?). Then, I'll check in with Hans and see if they have any opinions on the matter.

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: Important correction: `flutter build macos --flavor` does work. I was accidentally doing `flutter run -d macos --flavor` instead. Still, the other two points are valid, so I will file an issue and get Hans' input.

[the remark this answers is not in the document]
Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:584]: I had some offline conversations about this, so I'm adding an update here for posterity's sake. I'm allotting myself some time to investigate https://github.com/flutter/flutter/issues/139774. If getting flavors for macOS back under test proves simple enough, I'll do that and get support for it publicly documented.

File: packages/flutter_tools/test/commands.shard/hermetic/run_test.dart [hunk 3]
```diff
@@ -431,6 +431,32 @@ void main() {
...
+          'run',
+          '--no-pub',
+          '--no-hot',
+          '--flavor=vanilla'
+        ];
+
+        // Useful for test readability.
+        // ignore: avoid_redundant_argument_values
+        final FakeDevice device = FakeDevice(supportsFlavors: false);
+        testDeviceManager.devices = <Device>[device];
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/run_test.dart:445]: can we have two devices, one of which IS supported, and another that is NOT supported, and then pass the `--all` flag? I'm guessing in that case we should throw tool exit, right?

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 4]
```diff
@@ -610,6 +610,13 @@ class RunCommand extends RunCommandBase {
     webMode = featureFlags.isWebEnabled &&
       devices!.length == 1  &&
       await devices!.single.targetPlatform == TargetPlatform.web_javascript;
+
+    final String? flavor = stringArg('flavor');
+    final bool flavorsSupportedOnEveryDevice = devices!
+      .every((Device device) => device.supportsFlavors);
+    if (flavor != null && !flavorsSupportedOnEveryDevice) {
+      throwToolExit('--flavor is only supported for Android, macOS, and iOS devices.');
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/run.dart:618]: A few thoughts:

1. The API here of `List<Device>? devices` here is unfortunate, but I won't ask you to refactor this in this PR. I'm guessing `late final Future<List<Devices>> devices = findAllTargetDevices();` would probably be more readable (although it's unfortunate that `findAllTargetDevices()` has hidden dependencies, but alas)
2. I'm guessing that here `devices` will be only devices that: are available AND we have determined we will be running the app on, right? In which case, it seems like this validation logic is correct.
3. Do we also need validation logic for `flutter test`

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:618]: > 2\. I'm guessing that here `devices` will be only devices that: are available AND we have determined we will be running the app on, right? In which case, it seems like this validation logic is correct.

This was my understanding from the last time I read the code. It also makes sense with the name of `findAllTargetDevices`.

> 3\. Do we also need validation logic for flutter test

This seems like a good idea. I first need to figure out how device selection works for `flutter test`.

Comment (andrewkolos) [reply packages/flutter_tools/lib/src/commands/run.dart:618]: I've added the validation for `flutter test`. It is unclear to me whether or not this validation should itself be under test. Given its simplicity and that most of the neighboring `throwToolExits` also do not seem to be under test, I've elected to not create an integration test for this (perhaps I'm succumbing to broken window theory here). I did manually test this.

**Review by jmagman — APPROVED** [review]
The iOS and macOS parts LGTM, I'll let @christopherfujino finish his review.  

Thank you _so_ much for tracking down the macOS flavor issues.  Just today I was [able to implement something in packages tooling](https://github.com/flutter/packages/pull/5804#issuecomment-1877828475) for iOS and macOS and it's so much nicer when the tool works the same for both!

**Review by christopherfujino — APPROVED** [review]
lgtm

**Comment by mauriziopinotti** [comment]
What's the sense of this change? Now developers have to duplicate all run configurations, one for Web (without flavor option) and one for Mobile (with flavor option), causing massive inconvenience for no gain!

**Comment by andrewkolos** [comment]
> What's the sense of this change? Now developers have to duplicate all run configurations, one for Web (without flavor option) and one for Mobile (with flavor option), causing massive inconvenience for no gain!

When questioning the purpose of a PR, I recommend first reading both the PR and the issue associated with it. One or both of these should include the reason for the change. In this case, the original issue, https://github.com/flutter/flutter/issues/134197, includes the reasoning.

Regardless, a new issue has been opened to discuss this further, https://github.com/flutter/flutter/issues/143574.

**Comment by josehigroup** [comment]
It would be excellent if we could continue using the flavor flag without breaking the build. that way we can continue using our run configuration for mobile and web.

**Pull request merged**